src/extension/type_scheme.rs

```diff
@@ -0,0 +1,269 @@
+                        .unwrap()
+                )],
+                vec![USIZE_T]
+            )
+        );
+
+        Ok(())
+    }
+
+    fn id_fn(t: Type) -> FunctionType {
+        FunctionType::new(vec![t.clone()], vec![t])
+    }
+
+    #[test]
+    fn test_mismatched_args() -> Result<(), SignatureError> {
+        let ar_def = PRELUDE.get_type("array").unwrap();
+        let typarams = [TypeParam::Type(TypeBound::Any), TypeParam::max_nat()];
+        let tyvar = TypeArg::use_var(0, typarams[0].clone());
+        let szvar = TypeArg::use_var(1, typarams[1].clone());
+
+        // Valid schema...
+        let good_array =
+            Type::new_extension(ar_def.instantiate_concrete([tyvar.clone(), szvar.clone()])?);
+        let good_ts = OpDefTypeScheme::new(typarams.clone(), id_fn(good_array), &PRELUDE_REGISTRY)?;
+
+        // Sanity check (good args)
+        good_ts.compute_signature(
+            &[TypeArg::Type { ty: USIZE_T }, TypeArg::BoundedNat { n: 5 }],
+            &PRELUDE_REGISTRY,
+        )?;
+
+        let wrong_args = good_ts.compute_signature(
+            &[TypeArg::BoundedNat { n: 5 }, TypeArg::Type { ty: USIZE_T }],
+            &PRELUDE_REGISTRY,
+        );
+        assert_eq!(
+            wrong_args,
+            Err(SignatureError::TypeArgMismatch(
+                TypeArgError::TypeMismatch {
+                    param: typarams[0].clone(),
+                    arg: TypeArg::BoundedNat { n: 5 }
+                }
+            ))
+        );
+
+        // (Try to) make a schema with bad args
+        let arg_err = SignatureError::TypeArgMismatch(TypeArgError::TypeMismatch {
+            param: typarams[0].clone(),
+            arg: szvar.clone(),
+        });
+        assert_eq!(
+            ar_def.instantiate_concrete([szvar.clone(), tyvar.clone()]),
+            Err(arg_err.clone())
+        );
+        // ok, so that doesn't work - well, it shouldn't! So let's say we just have this signature (with bad args)...
+        let bad_array = Type::new_extension(CustomType::new(
+            ar_def.name().clone(),
+            [szvar, tyvar],
+            PRELUDE_ID,
+            TypeBound::Any,
+        ));
+        let bad_ts = OpDefTypeScheme::new(typarams.clone(), id_fn(bad_array), &PRELUDE_REGISTRY);
+        assert_eq!(bad_ts.err(), Some(arg_err));
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_misused_variables() -> Result<(), SignatureError> {
+        // Variables in args have different bounds from variable declaration
+        let tv = TypeArg::use_var(0, TypeParam::Type(TypeBound::Copyable));
+        let list_def = EXTENSION.get_type(&LIST_TYPENAME).unwrap();
+        let body_type = id_fn(Type::new_extension(list_def.instantiate_concrete([tv])?));
+        let reg = [EXTENSION.to_owned()].into();
+        for decl in [
+            TypeParam::Extensions,
+            TypeParam::List(Box::new(TypeParam::max_nat())),
+            TypeParam::Type(TypeBound::Any),
```

Comment:
Might as well include the other constructors of `TypeParam`

Reply:
Yes, fair, done. But I've removed some that are redundant given the more-detailed checks in `test_bound_covariance`